src/Authenticator/TokenAuthenticator.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         4.0.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Authenticator;
+
+use Authentication\Result;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Token Authenticator
+ *
+ * Authenticates an identity based on a token in a query param or the header.
+ */
+class TokenAuthenticator extends AbstractAuthenticator
+{
+
+    /**
+     * Checks if the token is in the headers or a request parameter
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request that contains login information.
+     * @return string|null
+     */
+    protected function _getToken(ServerRequestInterface $request)
+    {
+        $header = $this->config('header');
+        if (!empty($header)) {
+            return $request->getHeaderLine($header);
+        }
+
+        $param = $this->config('queryParam');
+        $queryParams = $request->getQueryParams();
+        if (!empty($queryParams[$param])) {
```

Comment:
Isset will return true even if it's an empty string or false. Empty seems more useful to me in this case?

> Determine if a variable is set and is not NULL. 

http://php.net/manual/en/function.isset.php